### ISSUE TITLE
Adding layers while using transfer training

### ISSUE THREAD

**Issue by Jessica-hub** [ultralytics/ultralytics#9537]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

If I added a block or layer at the beginning of the model in yolov8-obb.yaml. can I still load yolov8-obb.pt as pre-trained model?

**Comment by glenn-jocher** [comment]
@Jessica-hub sure! When adding a block or layer to the beginning of your model configuration (`yolov8-obb.yaml`), you can still utilize `yolov8-obb.pt` as a pre-trained model. However, keep in mind that layer weights for newly added layers won't be initialized from the pre-trained model - only the subsequent layers that match with the pre-trained model structure will have their weights loaded. 

You might need to adjust your training hyperparameters slightly to accommodate the learning needs of the newly added layers. A common approach is to use a lower learning rate initially or freeze the pre-trained layers for a few epochs. 

Happy coding! 😊

**Comment by Jessica-hub** [comment]
Thank you for your clear and precise answer. Can you let me know how to set up the hyperparameters to freeze the layers for certain epochs?

**Comment by glenn-jocher** [comment]
@Jessica-hub absolutely, happy to help! 😊 Freezing layers for a certain number of epochs isn't directly configurable through our current command-line arguments, but you can achieve this with a small modification to the training loop in the source code.

One approach is to modify the training script to:

1. Set the `requires_grad` attribute of the parameters you wish to freeze to `False` for the initial epochs.
2. After a certain number of epochs, set `requires_grad` back to `True` to unfreeze them.

Here's a quick snippet on how you might do it:

```python
# Assuming 'model' is your loaded YOLOv8 model
if epoch < freeze_epochs:
    for name, param in model.named_parameters():
        # Replace 'layer_to_freeze' with the actual layer names you wish to freeze
        if 'layer_to_freeze' in name: 
            param.requires_grad = False
else:
    for name, param in model.named_parameters():
        param.requires_grad = True
```

Remember to check the layer names in your model to accurately freeze the right ones! Hope this helps, and feel free to reach out with more questions.

**Comment by 233AvavAvA** [comment]
After increasing the number of layers, how does YOLOv8 know how the pre-training weights correspond to the layers in the yaml file?

**Comment by glenn-jocher** [comment]
Hey there! Great question! 😊 When you add layers to your model in the YAML file and use a pre-trained model like `yolov8-obb.pt`, YOLOv8 automatically maps the pre-trained weights to the corresponding layers based on names and shapes of the layers. If new layers are added or the architecture is modified, those layers simply won't have pre-trained weights and will need to be trained from scratch. The rest of the model benefits from the pre-training as usual. This seamless mixing of pre-trained and freshly initialized layers allows for flexible model customization. Hope this helps, and happy modeling!

**Comment by wyfshr** [comment]
In other words, no matter which layer of the network structure you add a module, you can load the pre-trained model yolov8n.pt normally, and only the newly added demon needs to be trained from scratch, right?

**Comment by glenn-jocher** [comment]
@wyfshr that's correct! When you add new layers to your YOLOv8 model configuration, YOLOv8 automatically aligns weights from the `yolov8n.pt` for the existing matched layers. Any new layers or modules do not initially have pre-trained weights and will start their training from scratch. This helps integrate the benefits of pre-training while allowing you to customize the architecture as needed. Happy coding! 😊
